nucypher/crypto/kits.py

```diff
@@ -55,11 +54,55 @@ def __init__(self,
                  ciphertext=None,
                  signature=NOT_SIGNED) -> None:
 
+        if type(capsule) == bytes:
+            raise Exception
+
         self.ciphertext = ciphertext
         self.capsule = capsule
         self.sender_verifying_key = sender_verifying_key
         self._signature = signature
 
+        self._cfrags = set()
+        self._delegating_key = None
+        self._receiving_key = None
+        self._verifying_key = None
+
+    def attach_cfrag(self, cfrag: VerifiedCapsuleFrag):
+        # TODO: check that the cfrag belongs to this capsule?
+        self._cfrags.add(cfrag)
+
+    def clear_cfrags(self):
+        self._cfrags = set()
+
+    def set_correctness_keys(self, delegating=None, receiving=None, verifying=None):
+
+        if self._delegating_key is None:
+            self._delegating_key = delegating
+        elif delegating is not None and delegating != self._delegating_key:
+            raise Exception("Replacing an existing delegating key")
```

Comment:
Similar to above, might benefit from a fancier exception class, but not a big deal.  Might warrant a TODO.

Reply:
Added a TODO for #2028. These are also sanity checks and can be avoided with a proper API design (so that one simply cannot re-assign the keys)